test/e2e/mtls_test.go

```diff
@@ -0,0 +1,180 @@
+	"k8s.io/apimachinery/pkg/util/intstr"
+	pkgTest "knative.dev/pkg/test"
+	"knative.dev/pkg/test/logstream"
+	"knative.dev/serving/pkg/apis/networking"
+	rtesting "knative.dev/serving/pkg/testing/v1"
+	"knative.dev/serving/test"
+	"knative.dev/serving/test/conformance/ingress"
+	v1test "knative.dev/serving/test/v1"
+)
+
+func testSvcToKubeSvc(t *testing.T, clients *test.Clients, injectA, injectB bool, expStatus int) {
+	t.Log("Creating a Pod and Service for the test app.")
+	name := test.ObjectNameForTest(t)
+	cleanup := createPodAndService(t, clients, name, injectA)
+	defer cleanup()
+
+	// Create envVars to be used in httpproxy app.
+	envVars := []corev1.EnvVar{{
+		Name:  targetHostEnv,
+		Value: name + "." + test.ServingNamespace + ".svc",
+	}}
+
+	t.Log("Creating a Knative Service for the httpproxy test app.")
+	names := test.ResourceNames{
+		Service: test.ObjectNameForTest(t),
+		Image:   "httpproxy",
+	}
+
+	test.CleanupOnInterrupt(func() { test.TearDown(clients, names) })
+	defer test.TearDown(clients, names)
+
+	resources, err := v1test.CreateServiceReady(t, clients, &names,
+		rtesting.WithEnv(envVars...),
+		rtesting.WithConfigAnnotations(map[string]string{
+			"sidecar.istio.io/inject": strconv.FormatBool(injectB),
+		}))
+	if err != nil {
+		t.Fatalf("Failed to create a Service: %v: %v", names.Service, err)
+	}
+
+	url := resources.Route.Status.URL.URL()
+	if _, err = pkgTest.WaitForEndpointState(
+		clients.KubeClient,
+		t.Logf,
+		url,
+		v1test.RetryingRouteInconsistency(pkgTest.MatchesAllOf(pkgTest.IsOneOfStatusCodes(expStatus))),
+		"HTTPProxy",
+		test.ServingFlags.ResolvableDomain,
+		test.AddRootCAtoTransport(t.Logf, clients, test.ServingFlags.Https),
+	); err != nil {
+		t.Fatalf("Get unexpected error: %v", err)
+	}
+
+}
+
+// In this test, we set up two apps: runtime and httpproxy.
+// runtime is a raw k8s pod and svc, while httpproxy is a knative service.
+// httpproxy is a proxy that redirects request to the runtime app.
+// The expected result is that the request sent to httpproxy app is successfully redirected
+// to runtime app when sidecar injection are both enabled or neither.
+func TestSvcToKubeSvc(t *testing.T) {
+	// testcases for table-driven testing.
+	var testCases = []struct {
+		name string
+		// injectA indicates whether istio sidecar injection is enabled for runtime (k8s) service.
+		// injectB indicates whether istio sidecar injection is enabled for httpproxy (knative) service.
+		injectA   bool
+		injectB   bool
+		expStatus int
+	}{
+		{"both-disabled", false, false, http.StatusOK},
+		{"a-disabled", false, true, http.StatusServiceUnavailable},
+		{"b-disabled", true, false, http.StatusBadGateway},
+		{"both-enabled", true, true, http.StatusOK},
+	}
+
+	t.Parallel()
+	cancel := logstream.Start(t)
+	defer cancel()
+
+	clients := Setup(t)
+
+	for _, tc := range testCases {
+		t.Run(tc.name, func(t *testing.T) {
+			if os.Getenv("MESH") != "1" && (tc.injectA || tc.injectB) {
```

Comment:
Do the inject tests actually fail in non-mesh?

Reply:
No, the inject tests pass all tests. But the all four tests(`(inject,inject)`,`(inject,no-inject)`, `(no-inject,inject)`,`(no-inject,no-inject)`) are same with `(no-inject,no-inject)` on no-mesh env, so I skipped three of them for short cut..